File: video/pipeline/lib/tts-edge.mjs
// Node wrapper around tts-edge.py — Microsoft Edge's free neural TTS service
// (the same "more human" voices Edge Reader uses). No API key, no account;
// requires internet access and the `edge_tts` Python package (see tts-edge.py).
// Quality tier sits between Windows SAPI (robotic, but offline) and ElevenLabs
// (premium, needs a key) — see tts.mjs for how the dispatcher picks engines.
import { spawn, spawnSync } from "node:child_process";
import { writeFileSync, readFileSync, mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { readAudioDurationMs, rescaleCues } from "./wav.mjs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PY = join(__dirname, "tts-edge.py");

export const DEFAULT_EDGE_VOICE = "en-US-EmmaMultilingualNeural";

const PYTHON_COMMANDS = ["python", "py"];

let pythonCmdCache = null;

// Windows can ship a Microsoft-Store "python" alias that opens the Store
// instead of running anything; probe each candidate with a real version
// check and pick the first one that actually works.
function findPythonCommand() {
  if (pythonCmdCache) return pythonCmdCache;
  for (const cmd of PYTHON_COMMANDS) {
    try {
      const probe = spawnSync(cmd, ["-c", "import edge_tts"], { windowsHide: true, timeout: 15000, encoding: "utf8" });
      if (probe.status === 0) {
        pythonCmdCache = cmd;
        return cmd;
      }
    } catch (err) {
      // candidate not found — try the next one
    }
  }
  pythonCmdCache = null;
  return null;
}

export function isEdgeTtsAvailable() {
  return Boolean(findPythonCommand());
}

function runPython(args) {
  const cmd = findPythonCommand();
  if (!cmd) return Promise.reject(new Error("Python with edge-tts not available"));
  return new Promise((resolve, reject) => {
    const proc = spawn(cmd, ["-u", PY, ...args], { windowsHide: true });
    let stderr = "";
    proc.stderr.on("data", (d) => (stderr += d.toString()));
    proc.on("error", reject);
    proc.on("close", (code) => {
      if (code !== 0) reject(new Error(`tts-edge.py exited ${code}: ${stderr.trim().slice(-2000)}`));
      else resolve();
    });
  });
}

// The pipeline's rate knob is Windows SAPI's -10..10 scale (shared with the
// SAPI path); edge-tts takes a percentage string. 4%/unit is a reasonable
// mapping (0 -> "+0%", 10 -> "+40%").
export function sapiRateToPercent(rate) {
  const clamped = Math.max(-10, Math.min(10, Number(rate) || 0));
  const pct = Math.round(clamped * 4);
  return `${pct >= 0 ? "+" : ""}${pct}%`;
}

export function edgeWordsToCues(words) {
  return (Array.isArray(words) ? words : [])
    .map((w) => ({ text: String(w.text ?? "").trim(), startMs: Number(w.startMs) || 0, endMs: Number(w.endMs) || 0 }))
    .filter((c) => c.text.length > 0 && c.endMs > 0);
}

export async function synthesizeWithEdge({ text, outAudioPath, voice, rate = 0, pitch = "+0Hz", volume = "+0%" }) {
  const tmp = mkdtempSync(join(tmpdir(), "sf-video-edge-"));
  const textPath = join(tmp, "text.txt");
  const timingPath = join(tmp, "timing.json");
  writeFileSync(textPath, text, "utf8");

  try {
    await runPython([
      "--text-path", textPath,
      "--out", outAudioPath,
      "--timing", timingPath,
      "--voice", voice || DEFAULT_EDGE_VOICE,
      "--rate", sapiRateToPercent(rate),
      "--pitch", pitch,
      "--volume", volume,
    ]);
    // Python's JSON has no BOM, but be defensive about UTF-8 BOM anyway.
    let rawText = readFileSync(timingPath, "utf8");
    if (rawText.charCodeAt(0) === 0xfeff) rawText = rawText.slice(1);
    const words = JSON.parse(rawText);
    return { rawWords: Array.isArray(words) ? words : [] };
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }
}

// Full voiceover + word-synced cues for the dispatcher in tts.mjs. Rescales
// the (already-cumulative) word timestamps to the MP3's measured duration so
// the last caption doesn't get cut off and the timeline matches real audio.
export async function edgeVoiceover({ text, outDir, jobId, voice, rate = 0, pitch = "+0Hz" }) {
  const audioPath = `${outDir}/${jobId}-vo.mp3`;
  const { rawWords } = await synthesizeWithEdge({ text, outAudioPath: audioPath, voice, rate, pitch });
  const rawCues = edgeWordsToCues(rawWords);
  const actualDurationMs = await readAudioDurationMs(audioPath);
  const cues = rescaleCues(rawCues, actualDurationMs);
  return { provider: "edge", audioPath, cues };
}

File: video/pipeline/lib/tts-edge.py
#!/usr/bin/env python3
# Synthesizes speech with Microsoft Edge's neural TTS service (the same free
# voices Edge Reader / edge-tts use — "more human" than the legacy Windows
# SAPI engine), writes the audio to a 24kHz mono MP3, and dumps exact
# word-level timestamps to a JSON file for caption sync.
#
# Requires the `edge_tts` Python package (free, no API key): `pip install
# edge-tts`. Node side calls this via pipeline/lib/tts-edge.mjs, which also
# handles falling back to Windows SAPI when Edge/the network isn't available.
#
# Why this exists instead of calling `edge-tts` CLI's --write-subtitles: the
# CLI writes *sentence*-level SRT cues; captions in this pipeline are word-
# synced, so we use the library's stream() API directly with
# boundary="WordBoundary" to get per-word offsets (which are cumulative
# across the whole utterance, even when long text gets split across
# requests). Offsets/durations arrive as 100ns ticks; we convert to ms here.
import argparse
import asyncio
import json
import sys


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--text-path", required=True)
    parser.add_argument("--out", required=True)
    parser.add_argument("--timing", required=True)
    parser.add_argument("--voice", default="en-US-EmmaMultilingualNeural")
    parser.add_argument("--rate", default="+0%")
    parser.add_argument("--pitch", default="+0Hz")
    parser.add_argument("--volume", default="+0%")
    return parser.parse_args()


async def synthesize(args):
    try:
        import edge_tts
    except ImportError as err:
        sys.stderr.write(
            "edge-tts Python package is not installed. Run: pip install edge-tts\n"
        )
        raise

    with open(args.text_path, encoding="utf-8") as handle:
        text = handle.read()

    communicate = edge_tts.Communicate(
        text,
        args.voice,
        rate=args.rate,
        pitch=args.pitch,
        volume=args.volume,
        boundary="WordBoundary",
    )

    words = []
    audio_bytes = 0
    with open(args.out, "wb") as media:
        async for chunk in communicate.stream():
            if chunk["type"] == "audio":
                media.write(chunk["data"])
                audio_bytes += len(chunk["data"])
            elif chunk["type"] == "WordBoundary":
                offset = chunk.get("offset") or 0
                duration = chunk.get("duration") or 0
                words.append(
                    {
                        "text": (chunk.get("text") or "").strip(),
                        "startMs": round(offset / 10000.0),
                        "endMs": round((offset + duration) / 10000.0),
                    }
                )

    if audio_bytes == 0:
        raise RuntimeError("Edge TTS returned no audio (is the network reachable?)")
    if not words:
        raise RuntimeError("Edge TTS returned no word boundaries")

    with open(args.timing, "w", encoding="utf-8") as handle:
        json.dump(words, handle, ensure_ascii=False)


def main():
    args = parse_args()
    try:
        asyncio.run(synthesize(args))
    except Exception as err:  # surface a clean, one-line failure to Node
        sys.stderr.write(f"edge-tts failed: {err}\n")
        sys.exit(1)


if __name__ == "__main__":
    main()
